content/blog/2019-09-25-floats-static-arrays.md

```diff
@@ -0,0 +1,67 @@
+
+These goals were mostly successful at the level of Bril semantics, parsing, and interpretation.  What was more difficult than expected (and less successful) were my attempts to model TypeScript values and arrays with [TS2Bril](https://github.com/sampsyo/bril/blob/master/bril-ts/ts2bril.ts).  Before exploring the limitations of the Bril representation, however, we must examine the floating point and array semantics of Bril more formally.
+
+### Floating Points in Bril
+
+Bril now supports the types `double` and `float`.  These represent, respectively,  the stadard IEEE-754 [double-precision](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) and [single-precision](https://en.wikipedia.org/wiki/Single-precision_floating-point_format) floating-point formats.  These types come equipped with the basic arithmetic operations `fadd`, `fsub`, `fmult`, and `fdiv`, along with the comparitives `feq`, `flt`, `fgt`, `fle`, and `fge`.  The introduction of these operations is intended to highlight that floating-point and integer operations should be treated as fundamentally different objects.
+
+In Bril, floating-point precisions should not be mixed.  It is expected, for instance, that the following code will produce a static exception:
+```
+v0: double = const 5;
+v1: float = const 5;
+v2: float = fadd v0 v1
+```
+Similarly, mixing floating-points and integers in Bril is undefined, as well as using floating-point arithmetic operations on integers or vice-versa.  In a future project, I intend to explore the consequences of relaxing these semantic requirements, particularly when mixing floating-point precision.
+
+### Bril Fixed-Length Arrays
+
+Bril now supports types of the form `array=type[size]`, where `type = | int | bool | float | double | array`.  Note the inclusion of type recursion in this definition, which permits multi-dimensional arrays.  Arrays in Bril must contain elements all of the same type, and the size of arrays cannot be changed after initialization.
+
+There are three operations related to manipulating arrays in Bril: initialization, assignment, and indexing.  Arrays are initialized with the `new` instruction, which takes a type and stores a default initialization of that type to the given destination.  Array elements can be set with the `set` instruction, an effectful instruction which takes the array, an index, and a value in that order.  Finally, array elements can be read using the `index` instruction, which writes a value of the given array element.
```

Comment:
For `new`, I assume that the type must be an array type? Or can it be any type?

Reply:
I know this is bikeshedding, but because one of the operations is `set`, I probably would have called the other one `get` instead of `index`. :smiley: